README.md
# aws_s3_upload

A simple, convenient package for uploading to S3.

_Heavily_ inspired by [this stackoverflow answer](https://stackoverflow.com/a/54983831/2330228)

## Getting Started

Having created credentials on AWS, upload a file like so:

```dart
AwsS3.uploadFile(
  accessKey: "AKxxxxxxxxxxxxx",
  secretKey: "xxxxxxxxxxxxxxxxxxxxxxxxxx",
  file: File("path_to_file"),
  bucket: "bucket_name",
  region: "us-east-2",
  metadata: {"test": "test"} // optional
);
```

## Warning

Before using this package, consider if it's acceptable to include AWS keys in your final app. Doing so might be okay for internal or trusted apps, but think twice before including keys in an app published on the app store.

## Motivation / Disclaimer

There already exists a number of Flutter plugins for interacting with S3, some of which are more actively maintained. This small library was built because the few I tried either failed to work out of the box, or required the use of a Pool ID and AWS Cognito, which my project doesn't use. YMMV.


CHANGELOG.md
## [1.5.0] - 1/22/2023

Support including metadata in uploads.

## [1.4.0] - 6/29/2022

Support setting object key directly.

## [1.3.0] - 6/29/2022

Allow setting ACL and fix endpoint.

## [1.2.2] - 2/2/2022

readded gitignore and uploaded properly

## [1.2.1] - 2/2/2022

Added enum for acl to be able to choose.

## [1.2.0] - 5/9/2021

Null safety and updated dependencies.

## [1.1.2] - 3/20/2021

Fix returned url when using custom filename.

## [1.1.0] - 3/20/2021

Add support for uploading with a filename different from the file's current filename.

## [1.0.2] - 3/20/2021

Update documentation to reflect class renaming.

## [1.0.1] - 3/20/2021

Remove leftover template placeholders and restructure more logically.

## [1.0.0] - 3/20/2021

Initial Release


aws_s3_upload-1.3.0/test/aws_s3_upload_test.dart
import 'package:flutter_test/flutter_test.dart';

// TODO: Write Tests(?) May be difficult without exposing credentials
void main() {
  test('check test run', () {
    expect(true, true);
  });
}


lib/src/policy.dart
import 'dart:convert';

import 'package:amazon_cognito_identity_dart_2/sig_v4.dart';
import 'package:aws_s3_upload/enum/acl.dart';
import 'package:aws_s3_upload/src/utils.dart';

class Policy {
  String expiration;
  String region;
  ACL acl;
  String bucket;
  String key;
  String credential;
  String datetime;
  int maxFileSize;
  Map<String, dynamic>? metadata;

  Policy(
    this.key,
    this.bucket,
    this.datetime,
    this.expiration,
    this.credential,
    this.maxFileSize,
    this.acl, {
    this.region = 'us-east-2',
    this.metadata,
  });

  factory Policy.fromS3PresignedPost(
    String key,
    String bucket,
    String accessKeyId,
    int expiryMinutes,
    int maxFileSize,
    ACL acl, {
    String region = 'us-east-2',
    Map<String, dynamic>? metadata,
  }) {
    final datetime = SigV4.generateDatetime();
    final expiration = (DateTime.now())
        .add(Duration(minutes: expiryMinutes))
        .toUtc()
        .toString()
        .split(' ')
        .join('T');
    final cred =
        '$accessKeyId/${SigV4.buildCredentialScope(datetime, region, 's3')}';

    return Policy(
      key,
      bucket,
      datetime,
      expiration,
      cred,
      maxFileSize,
      acl,
      region: region,
      metadata: metadata,
    );
  }

  String encode() {
    final bytes = utf8.encode(toString());
    return base64.encode(bytes);
  }

  List<Map<String, String>> _convertMetadataToPolicyParams(
      Map<String, dynamic>? metadata) {
    final List<Map<String, String>> params = [];

    if (metadata != null) {
      for (var k in metadata.keys) {
        params.add({k: metadata[k]});
      }
    }

    return params;
  }

  @override
  String toString() {
    final metadataParams = _convertMetadataToPolicyParams(metadata);

    final payload = {
      "expiration": "${this.expiration}",
      "conditions": [
        {"bucket": "${this.bucket}"},
        ["starts-with", "\$key", "${this.key}"],
        ["starts-with", "\$Content-Type", ""],
        {"acl": "${aclToString(acl)}"},
        ["content-length-range", 1, this.maxFileSize],
        {"x-amz-credential": "${this.credential}"},
        {"x-amz-algorithm": "AWS4-HMAC-SHA256"},
        {"x-amz-date": "${this.datetime}"},
      ],
    };

    // If there's metadata, add it to the list of conditions for the policy.
    if (metadataParams.isNotEmpty) {
      for (final p in metadataParams) {
        (payload['conditions'] as List).add(p);
      }
    }

    return jsonEncode(payload);
  }
}


lib/aws_s3_upload.dart
library aws_s3_upload;

import 'dart:io';

import 'package:amazon_cognito_identity_dart_2/sig_v4.dart';
import 'package:aws_s3_upload/enum/acl.dart';
import 'package:aws_s3_upload/src/utils.dart';
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as path;
import 'package:recase/recase.dart';

import './src/policy.dart';

/// Convenience class for uploading files to AWS S3
class AwsS3 {
  /// Upload a file, returning the file's public URL on success.
  static Future<String?> uploadFile({
    /// AWS access key
    required String accessKey,

    /// AWS secret key
    required String secretKey,

    /// The name of the S3 storage bucket to upload  to
    required String bucket,

    /// The file to upload
    required File file,

    /// The key to save this file as. Will override destDir and filename if set.
    String? key,

    /// The path to upload the file to (e.g. "uploads/public"). Defaults to the root "directory"
    String destDir = '',

    /// The AWS region. Must be formatted correctly, e.g. us-west-1
    String region = 'us-east-2',

    /// Access control list enables you to manage access to bucket and objects
    /// For more information visit [https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html]
    ACL acl = ACL.public_read,

    /// The filename to upload as. If null, defaults to the given file's current filename.
    String? filename,

    /// The content-type of file to upload. defaults to binary/octet-stream.
    String contentType = 'binary/octet-stream',

    /// Additional metadata to be attached to the upload
    Map<String, String>? metadata,
  }) async {
    final endpoint = 'https://$bucket.s3.$region.amazonaws.com';

    var uploadKey;

    if (key != null) {
      uploadKey = key;
    } else if (destDir.isNotEmpty) {
      uploadKey = '$destDir/${filename ?? path.basename(file.path)}';
    } else {
      uploadKey = '${filename ?? path.basename(file.path)}';
    }

    final stream = http.ByteStream(Stream.castFrom(file.openRead()));
    final length = await file.length();

    final uri = Uri.parse(endpoint);
    final req = http.MultipartRequest("POST", uri);
    final multipartFile = http.MultipartFile('file', stream, length,
        filename: path.basename(file.path));

    // Convert metadata to AWS-compliant params before generating the policy.
    final metadataParams = _convertMetadataToParams(metadata);

    // Generate pre-signed policy.
    final policy = Policy.fromS3PresignedPost(
      uploadKey,
      bucket,
      accessKey,
      15,
      length,
      acl,
      region: region,
      metadata: metadataParams,
    );

    final signingKey =
        SigV4.calculateSigningKey(secretKey, policy.datetime, region, 's3');
    final signature = SigV4.calculateSignature(signingKey, policy.encode());

    req.files.add(multipartFile);
    req.fields['key'] = policy.key;
    req.fields['acl'] = aclToString(acl);
    req.fields['X-Amz-Credential'] = policy.credential;
    req.fields['X-Amz-Algorithm'] = 'AWS4-HMAC-SHA256';
    req.fields['X-Amz-Date'] = policy.datetime;
    req.fields['Policy'] = policy.encode();
    req.fields['X-Amz-Signature'] = signature;
    req.fields['Content-Type'] = contentType;

    // If metadata isn't null, add metadata params to the request.
    if (metadata != null) {
      req.fields.addAll(metadataParams);
    }

    try {
      final res = await req.send();

      if (res.statusCode == 204) return '$endpoint/$uploadKey';
    } catch (e) {
      print('Failed to upload to AWS, with exception:');
      print(e);
      return null;
    }
  }

  /// A method to transform the map keys into the format compliant with AWS.
  /// AWS requires that each metadata param be sent as `x-amz-meta-*`.
  static Map<String, String> _convertMetadataToParams(
      Map<String, String>? metadata) {
    Map<String, String> updatedMetadata = {};

    if (metadata != null) {
      for (var k in metadata.keys) {
        updatedMetadata['x-amz-meta-${k.paramCase}'] = metadata[k]!;
      }
    }

    return updatedMetadata;
  }
}


lib/src/signer.dart
import 'dart:io';

import 'package:amazon_cognito_identity_dart_2/sig_v4.dart';
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as path;

import './policy.dart';

/// Convenience class for uploading files to AWS S3
class AwsS3 {
  /// Upload a file, returning the file's public URL on success.
  static Future<String> uploadFile({
    /// AWS access key
    String accessKey,

    /// AWS secret key
    String secretKey,

    /// The name of the S3 storage bucket to upload  to
    String bucket,

    /// The path to upload the file to (e.g. "uploads/public"). Defaults to the root "directory"
    String destDir,

    /// The AWS region. Must be formatted correctly, e.g. us-west-1
    String region = 'us-east-2',

    /// The file to upload
    File file,
  }) async {
    AwsS3.uploadFile(
        accessKey: "AKxxxxxxxxxxxxx",
        secretKey: "xxxxxxxxxxxxxxxxxxxxxxxxxx",
        file: File("path_to_file"),
        bucket: "bucket_name",
        region: "us-east-2");

    final endpoint = 'https://$bucket.s3-$region.amazonaws.com';

    final stream = http.ByteStream(Stream.castFrom(file.openRead()));
    final length = await file.length();

    final uri = Uri.parse(endpoint);
    final req = http.MultipartRequest("POST", uri);
    final multipartFile = http.MultipartFile('file', stream, length, filename: path.basename(file.path));

    final policy = Policy.fromS3PresignedPost('$destDir/${path.basename(file.path)}', bucket, accessKey, 15, length,
        region: region);
    final key = SigV4.calculateSigningKey(secretKey, policy.datetime, region, 's3');
    final signature = SigV4.calculateSignature(key, policy.encode());

    req.files.add(multipartFile);
    req.fields['key'] = policy.key;
    req.fields['acl'] = 'public-read';
    req.fields['X-Amz-Credential'] = policy.credential;
    req.fields['X-Amz-Algorithm'] = 'AWS4-HMAC-SHA256';
    req.fields['X-Amz-Date'] = policy.datetime;
    req.fields['Policy'] = policy.encode();
    req.fields['X-Amz-Signature'] = signature;

    try {
      final res = await req.send();

      if (res.statusCode == 204) return '$endpoint/$destDir/${path.basename(file.path)}';
    } catch (e) {
      print(e.toString());
    }
  }
}
